### PR TITLE
Move interfaces to one package avoiding go circular imports

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Move top level history service interface def to one package avoiding unnecessary go circular imports

**Why?**
See above

**How did you test it?**
N/A

**Potential risks**
N/A

**Is hotfix candidate?**
N/A

### CODE REVIEW

**Review by paulnpdev — APPROVED** [review]

File: service/history/definition/event_notifier.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+
+	enumsspb "go.temporal.io/server/api/enums/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/definition"
+)
+
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination event_notifier_mock.go
```
Comment (rodrigozhou) [inline service/history/definition/event_notifier.go:38]: Move this to right after the license, before the package name line.

File: service/history/definition/event_cache.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+import (
+	"context"
+
+	historypb "go.temporal.io/api/history/v1"
+
+	"go.temporal.io/server/common/namespace"
+)
+
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination event_cache_mock.go
```
Comment (rodrigozhou) [inline service/history/definition/event_cache.go:35]: Move this to right after the license, before the package name line.

File: service/history/definition/shard_context.go [hunk 3]
```diff
@@ -37,29 +37,27 @@ import (
...
 	"go.temporal.io/server/common/persistence"
 	"go.temporal.io/server/common/persistence/serialization"
 	"go.temporal.io/server/common/searchattribute"
 	"go.temporal.io/server/service/history/configs"
-	"go.temporal.io/server/service/history/events"
 	"go.temporal.io/server/service/history/tasks"
 )
 
-//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination context_mock.go
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination shard_context_mock.go
```
Comment (rodrigozhou) [inline service/history/definition/shard_context.go:50]: ditto

File: service/history/definition/shard_controller.go [hunk 4]
```diff
@@ -22,22 +22,22 @@
...
-package shard
+package definition
 
 import (
 	"go.temporal.io/server/common"
 	"go.temporal.io/server/common/namespace"
 )
 
-//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination controller_mock.go
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination shard_controller_mock.go
```
Comment (rodrigozhou) [inline service/history/definition/shard_controller.go:32]: ditto

File: service/history/definition/workflow_cache.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+import (
+	"context"
+
+	commonpb "go.temporal.io/api/common/v1"
+
+	"go.temporal.io/server/common/namespace"
+)
+
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination workflow_cache_mock.go
```
Comment (rodrigozhou) [inline service/history/definition/workflow_cache.go:35]: ditto

File: service/history/definition/workflow_context.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+	"go.temporal.io/server/common/definition"
+	"go.temporal.io/server/common/persistence"
+)
+
+const (
+	CallerTypeAPI  CallerType = 0
+	CallerTypeTask CallerType = 1
+)
+
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination workflow_context_mock.go
```
Comment (rodrigozhou) [inline service/history/definition/workflow_context.go:41]: ditto

**Review by rodrigozhou — APPROVED** [review]
Isn't `defs` more standard name for this instead of `definition`? No strong opinion though.

Left some nit comments in the code.

**Comment by wxing1292** [comment]
> Isn't `defs` more standard name for this instead of `definition`? No strong opinion though.

i have no opinions about `def` vs `definition`, what do you guys think?

**Review by mindaugasrukas — CHANGES_REQUESTED** [review]
Let's review this collectively. This change has a [code smell](https://en.wikipedia.org/wiki/Code_smell).

**Comment by mindaugasrukas** [comment]
@wxing1292, please provide details where are `unnecessary go circular imports`. And if possible, please provide a full context of the issue and why the proposed solution should be acceptable.

IMO, this change should deserve an extensive design review.

**Comment by wxing1292** [comment]
> Let's review this collectively. This change has a [code smell](https://en.wikipedia.org/wiki/Code_smell).

i do not believe it is a good idea to simply block someone's PR to learn something or figure out a code.

>please provide details where are unnecessary go circular imports

simply google search can provide some info, e.g. https://stackoverflow.com/questions/28256923/import-cycle-not-allowed

**Comment by mindaugasrukas** [comment]
Have we considered extracting the cache to a dedicated package?

**Comment by wxing1292** [comment]
> Have we considered extracting the cache to a dedicated package?

demo-ed offline, at least there is no easy solution right now.

**Comment by mindaugasrukas** [comment]
Would this work https://github.com/temporalio/temporal/pull/3601?

**Review by alexshtin — CHANGES_REQUESTED** [review]
I don't like this change. It goes against basic go principles. Yes, we probably can't refactor entire code base and move interfaces on caller side right now. Or even get rid of interfaces where they are not really needed. But this is move in opposite direction.

The fact that you need such kind of package (`common`, `util`, `helpers`, `definition`, etc) indicates that code is not structured properly. And by proper structure, I mean small packages with well defined scope and responsibility. And again, I understand, that you don't have time to do it right now.

What is your immediate need? Can you do a minimum small change instead? And let's discuss long term plan around `service/history` refactoring.

**Comment by alexshtin** [comment]
To support my point:
> Don’t use a single package for all your APIs. Many well-intentioned programmers put all the interfaces exposed by their program into a single package named `api`, `types`, or `interfaces`, thinking it makes it easier to find the entry points to their code base. This is a mistake. Such packages suffer from the same problems as those named `util` or `common`, growing without bound, providing no guidance to users, accumulating dependencies, and colliding with other imports. Break them up, perhaps using directories to separate public packages from implementation.

From: https://go.dev/blog/package-names.

**Comment by wxing1292** [comment]
>Yes, we probably can't refactor entire code base and move interfaces on caller side right now. 

can we ever?
how many callers for e.g. shard context? how many for mutable state? if we create the interface on the caller side, how many same / similar interfaces should we create & **manage** ? is this practical?

>Or even get rid of interfaces where they are not really needed.

name few interfaces which are not needed.
removing interfaces means making it harder to decoupling components, harder to write tests (no mocks)

>The fact that you need such kind of package (common, util, helpers, definition, etc) indicates that code is not structured properly. And by proper structure, I mean small packages with well defined scope and responsibility.

no idea why `common`, `util`, `helpers` are mentioned here, i am going to ignore.
the definition package is introduced here aiming to reduce the frequency of go circular imports (this issue is rather annoying)

imagine if you are a dev attempting to add a new feature

example 1:
before: package1 → package0
after: package1 → package0; `new logic: package0 → package1` ← import cycles

example 2:
before: package2 → package1; package1 → package0
after: package2 → package1; package1 → package0; `new logic: package0 → package2` ← import cycles

keeping the existing structure **may** still allow the open source server to survive for a long time, at the cost of dev continuing to moving logic around. 

what i proposed here (you should take a look at the internal doc) is to have a dedicated package for interface / struct definition, so concrete implementations only need to import this definition package. this decrease the possibility of go import cycles

>And again, I understand, that you don't have time to do it right now.

do you?

>What is your immediate need? Can you do a minimum small change instead? And let's discuss long term plan around service/history refactoring.

are you familiar with our persistence layer struct / interface definitions? if not, take a look. this PR pretty much does the same. 
if our persistence layer can do a decent job supporting Cassandra /  SQL (MySQL / PostgreSQL / SQLite) & other DBs use by our anonymous OSS users, then this PR should (begin to) allow more customizability as well 

>... thinking it makes it easier to find the entry points to their code base ...

this is not the intent, see above resp

**Comment by mindaugasrukas** [comment]
> Let's review this collectively. This change has a [code smell](https://en.wikipedia.org/wiki/Code_smell).

I publicly apologize for my unprofessional comment about this change having a code smell.
Also, I'm unblocking this PR and leaving the decision to the repository owners.

As an alternative to this change, please consider https://github.com/temporalio/temporal/pull/3601.

File: service/history/ndc/activity_replicator.go [hunk 7]
```diff
@@ -107,7 +106,7 @@ func (r *ActivityReplicatorImpl) SyncActivity(
 		ctx,
 		namespaceID,
 		execution,
-		workflow.CallerTypeAPI,
+		definition.CallerTypeTask,
```
Comment (yycptt) [inline service/history/ndc/activity_replicator.go:109]: `CallerTypeAPI`

File: service/history/ndc/history_replicator.go [hunk 8]
```diff
@@ -323,7 +322,7 @@ func (r *HistoryReplicatorImpl) applyEvents(
 		ctx,
 		task.getNamespaceID(),
 		*task.getExecution(),
-		workflow.CallerTypeAPI,
+		definition.CallerTypeTask,
```
Comment (yycptt) [inline service/history/ndc/history_replicator.go:325]: `CallerTypeAPI`

File: service/history/ndc/transaction_manager.go [hunk 9]
```diff
@@ -444,7 +443,7 @@ func (r *transactionMgrImpl) loadWorkflow(
 			WorkflowId: workflowID,
 			RunId:      runID,
 		},
-		workflow.CallerTypeAPI,
+		definition.CallerTypeTask,
```
Comment (yycptt) [inline service/history/ndc/transaction_manager.go:446]: `CallerTypeAPI`

**Review by yycptt — COMMENTED** [review]
There're a lot more in other files and tests. Plz double check.

**Comment by yiminc** [comment]
Close this PR as we are not doing this for now.

**Pull request closed without merging**